### DISCUSSION TITLE
SemanticTransformer training

### DISCUSSION THREAD

**Discussion by amitaie** [lucidrains/audiolm-pytorch#202] [Q&A]
Hey, anyone that succeeded in training an AudioLM model from this repo or maybe just the SemanticTransformer and is willing to share the training params, how long did you train? some other tips? 

I'm using 64 batch size, 2 sec for sample, and the other are default and i'm getting to loss ~2 very quickly but it seems that the model isn't getting better.

Thanks in advanced!

**Comment by LWprogramming** [comment]
I think before [this commit](https://github.com/lucidrains/audiolm-pytorch/commit/95724a344ae229dbaa65564cfd09e510f2945e21) I was also getting loss around 2 and not decreasing, so I might check that a) you have the latest version and b) you're able to intentionally overfit to a tiny dataset. If you still have loss stuck at 2 even on a tiny dataset, then that seems like a sign of a bug.

After the commit, I was able to take audiolm_pytorch, trained it to a very tiny dataset made up of the same data file copied a bunch of times, and was able to intentionally overfit it so the loss was near (or sometimes exactly) 0 and the output sounded pretty similar to the input!

Here's the original input and what I was able to generate using the overfitted network:

[result of overfitting](https://github.com/lucidrains/audiolm-pytorch/assets/13173037/787a3e20-d9fa-42a4-9c92-ef8263d5cffc)

[original sample](https://github.com/lucidrains/audiolm-pytorch/assets/13173037/76a87a8e-cf96-4442-85cd-5f1a789bb22a
)

Since the dataset was intentionally so tiny I didn't really need a ton of steps-- I set 5000 but the loss went to 0 really fast, probably unnecessary. Batch size/ accum grad every was 1 and I trained on a 1 second sample. The sample I used is at 24kHz, since I'm using EnCodec and wanted to avoid any potential issues in case there's any bug in the resampling part (rather than the model itself).

**Reply by lucidrains** [reply]
@LWprogramming i think that commit should only affect sampling and not training?

nonetheless, it seems like it is working for you :smile:

**Reply by LWprogramming** [reply]
Oops, yes, it's the version I trained on to make this, but the actual change that this run has (that previous runs didn't) that probably fixed the training was the encodec fix a99b26c49afb7c3eb59767ce4a708305fff81e21

**Reply by lucidrains** [reply]
oh yes, makes sense!

**Reply by amitaie** [reply]
@LWprogramming Thanks for the answer, I'll try doing that. 
In the paper they used different hyper-params, such as 12 layers, 16 heads and etc. Did you try different hyper-params? different length of inputs?

You said oy used batch size =1, i'm assuming only for the dummy batch? any ideas what is needed for a standard training?

**Reply by LWprogramming** [reply]
Yes, for the dummy batch-- I'm going to try batch 8 grad accum 16 on a full dataset, but I don't know for sure that it works. Haven't touched hyperparams at all haha, do you know if it tends to make a difference?

**Reply by amitaie** [reply]
@LWprogramming I have no idea, tring some stuff now but I think I have some bug in the training because i'm getting loss like this:

![image](https://github.com/lucidrains/audiolm-pytorch/assets/113421133/2e30d9a9-5c87-4b7b-b620-1518ec09d2bb)

Saw something like that? anyone?

**Reply by LWprogramming** [reply]
all my training scripts are [here](https://github.com/LWprogramming/audiolm-pytorch-training/) if you want to see if my code (which seems to work) works on your machine

**Reply by amitaie** [reply]
@LWprogramming  I'll look at it, thanks!
Can you share your's semantic loss graph? or just the range that the loss was?

**Reply by amitaie** [reply]
@LWprogramming I think i found my bug, I somehow didn't use grad_clip. Nevertheless my accuracy in the Semantic transformer is around 0.5, do you (or anyone else) got to a higher results? 0.5 sounds to me too low..

**Reply by amitaie** [reply]
@LWprogramming Did you tried training on a real dataset? I'm training Semantic and Coarse, the loss is going down and everything fine but when i'm generating there is no reasonable speech, mainly bla bla. Can youshare some of your results?

my Semantic loss after 140K steps of batch size 64 and data len of 320 * 100 is ~1.76
my Coarse loss after 300K steps of batch size 32 and data len of 320 * 100 is ~2.82

**Reply by LWprogramming** [reply]
Hi! I've been trying to get multi-gpu working but keep running into bugs so no results yet :(

**Comment by acul3** [comment]
sorry for jumping in

about semantic transformers, if i want to train language with non-english,should i train hubert/wav2vec with that language first?
since we need hubert/wav2vec checkpoint to train semantic?

**Comment by amitaie** [comment]
On the same subject but different angle - what wav2vec did you use? i'm using Hubert with 500 clustering but in the paper they used  w2v-bert with 1000 clustering. anyone trained something with 1000 clusters?

**Reply by LWprogramming** [reply]
I'm using [500](https://github.com/LWprogramming/audiolm-pytorch-training/blob/17fa99164d8dd1ddf0eb74ff48038334826531c4/audiolm_pytorch_demo_laion.py#L142) as well and it seems to be working so far
